File: 3286-find-a-safe-walk-through-a-grid/3286-find-a-safe-walk-through-a-grid.java
class Solution {
    boolean[][] vis;
    int[][][] dp;
    boolean check(List<List<Integer>> grid, int row, int col, int hlt) {
        int m = grid.size(); int n = grid.get(0).size();
        if (row >= m || col >= n || row<0 || col<0 || hlt<=0 || vis[row][col]) return false;
        if (row==m-1 && col==n-1) {
            return hlt>grid.get(row).get(col);
        }
        if (dp[row][col][hlt]!=-1) return dp[row][col][hlt]==1;
        vis[row][col] = true;
        boolean up = check(grid,row-1,col,hlt-grid.get(row).get(col));
        boolean dw = check(grid,row+1,col,hlt-grid.get(row).get(col));
        boolean lf = check(grid,row,col-1,hlt-grid.get(row).get(col));
        boolean rf = check(grid,row,col+1,hlt-grid.get(row).get(col));
        vis[row][col] = false;
        boolean result = up || dw || lf || rf;
        if(result){
            dp[row][col][hlt]=1;
        }
        else dp[row][col][hlt]=0;
        return result;
    }
    public boolean findSafeWalk(List<List<Integer>> grid, int health) {
        int m = grid.size();
        int n = grid.get(0).size();
        dp = new int[m][n][health+1];
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                Arrays.fill(dp[i][j],-1);
            }
        }
        vis = new boolean[m][n];
        return check(grid, 0, 0, health);
    }
}

File: 3286-find-a-safe-walk-through-a-grid/3286-find-a-safe-walk-through-a-grid.cpp
class Solution {
public:
bool findSafeWalk(vector<vector<int>>& g, int health) {
    int m = g.size(), n = g[0].size(), cnt[51][51] = {};
    priority_queue<array<int, 3>> pq;
    pq.push({health - g[0][0], 0, 0});
    while (!pq.empty()) {
        auto [h, i, j] = pq.top(); pq.pop();
        if (i == m - 1 && j == n - 1)
            return true;
        for (auto [dx, dy] : vector<pair<int, int>>{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}) {
            int x = i + dx, y = j + dy;
            if (min(x, y) >= 0 && x < m && y < n && cnt[x][y] < h - g[x][y]) {
                cnt[x][y] = h - g[x][y];
                pq.push({h - g[x][y], x, y});
            }
        }
    }
    return false;
}
};
